opt_einsum/path_random.py

```diff
@@ -0,0 +1,320 @@
+import math
+import time
+import heapq
+import random
+import functools
+from collections import deque
+
+from . import helpers
+from .paths import PathOptimizer, _ssa_optimize, _BETTER_FNS, ssa_to_linear, _calc_k12_flops
+
+
+__all__ = ["RandomGreedy", "random_greedy"]
+
+
+class RandomOptimizer(PathOptimizer):
+    """Base class for running any random path finder that benefits
+    from repeated calling, possibly in a parallel fashion. Custom random
+    optimizers should subclass this, and the ``setup`` method should be
+    implemented with the following signature::
+
+        def setup(self, inputs, output, size_dict):
+            # custom preparation here ...
+            return trial_fn, trial_args
+
+    Where ``trial_fn`` itself should have the signature::
+
+        def trial_fn(r, *trial_args):
+            # custom computation of path here
+            return ssa_path, cost, size
+
+    Where ``r`` is the run number and could for example be used to seed a
+    random number generator. See ``RandomGreedy`` for an example.
+
+
+    Parameters
+    ----------
+    max_repeats : int, optional
+        The maximum number of repeat trials to have.
+    max_time : float, optional
+        The maximum amount of time to run the algorithm for.
+    minimize : {'flops', 'size'}, optional
+        Whether to favour paths that minimize the total estimated flop-count or
+        the size of the largest intermediate created.
+    executor : executor-pool like, optional
```

Comment:
Could `executor` be more like:
- `threading=True` which would build a ThreadPool
- `int` which would pick the number of threads to use
- `ThreadPool` to use

Might make it a bit easier on the user.

Reply:
Yes, I could definitely add ``executor=True`` and ``executor=int`` options, such that a ``concurrent.futures.ProcessPoolExecutor`` is created automatically with either the default or specified number of workers. I feel like ``executor`` is the most precise name, but we could also think about changing the keyword to ``parallel`` or something?

Reply:
What about `nthread`? I suspect that most folks would let the function build its own executor.

Reply:
Given above comments about threading, maybe ``parallel`` works best as a compromise? I think the three possible use cases all work reasonably well:
```python
RandomGreedy(parallel=True)  # probably the most used
RandomGreedy(parallel=4)
RandomGreedy(parallel=pool)
```
whereas something like ``nproc=pool`` might be a bit confusing.

Reply:
Sounds good and I agree that `parallel=True` is likely the most common case here.